### ISSUE TITLE
Proposal: define default for all

### ISSUE THREAD

**Issue by laukstein** [WICG/feature-policy#189]
I propose `Feature-Policy` adopt `Content-Security-Policy` concept and be able to define all default, example:

    Feature-Policy: default 'none'
    Feature-Policy: default 'none'; fullscreen 'self'

where similarity in CSP would be

    Content-Security-Policy: default-src 'none'
    Content-Security-Policy: default-src 'none'; img-src 'self'

Otherwise we would be stuck to define all features separately, and since the features list will grow more and more, without the global default it likely will harmful web.

**Comment by CharlesStover** [comment]
I second this, and came here for the same suggestion. I would prefer my features be blacklisted by default and whitelisted as needed. My Feature-Policy header is ridiculously long when I have to explicitly define all of them as `'none'`.

I also fear that as time progresses, browsers and devices will add non-standardized features to their feature lists. It will become nearly impossible to blacklist all non-standardized features.

**Comment by nico3333fr** [comment]
Same suggestion for me: it could be useful to define a default to `'none'` and then define what we need on the website/app.

**Comment by eeeps** [comment]
Just a note to say that this relates to some of the discussion here: https://github.com/WICG/feature-policy/issues/129, where a default for a specific subset of policy controlled features (client hints) was proposed.

**Comment by olmari** [comment]
Definately thumbs up for at least possibility of defining default policy, having a (growing) blacklist with no default is insanity.. Even if some feature gets added and then a site with default policy breaks, it is matter of allowing that one policy.. In current form it is impossible to manage every possible new addition by hand.

**Comment by ojanvafai** [comment]
We already have some experience with this with iframe sandbox where we are having difficulty adding new directives because they break pages. With FeaturePolicy, it would be a bit different because we can add new directives without having them be part of the default policy, but it's still hard for me to see how we would be able to add new directives over time to the default policy. With iframe sandbox we can at least prioritize it due to the security benefits.

It's difficult for browsers to ship things that break web sites, especially since many sites are unmaintained, so the ease of sites fixing when we break them doesn't really mitigate that concern.

**Comment by olmari** [comment]
@ojanvafai (and all whom may be concerned) AFAIK What we (at least me) are wanting to have an way to _tell_ desired default policy/policies. So without any default defined standard could very well be "opt-in", but **if** header has something like _default 'none'_ **then** honour that and deny everything if not told otherwise..

This way sites not having this header at all will continue to work as is today, and administrators who want's to just blacklist certain things regardless of new possible directives can do so, but it would also allow most concerned admins to introduce an default policy and then whitelist allowed ones.

**Comment by olmari** [comment]
@clelland then you don't just use "default none" policy... I don't know much more clearer I can be on my proposion... Having written no policy would be same as current behaviour. having written default none policy would do just that... All admin configurable...

**Comment by clelland** [comment]
@olmari, I realize that you mean for there to be no behaviour changes without explicitly setting the default. The problem that I see is that setting such a default on your site is necessarily a huge risk -- your site would be at far more risk of breaking with future browser revisions than has ever been the case in the past.

Although that's probably not the reality of the situation -- in reality, the existence of such an option, especially if it had any significant level of adoption on the web, would mean that browser vendors would be effectively barred from introducing any new features which would break content on all of those sites. Every new feature decision would have to include serious consideration and discussion about the impact to all of the sites which use `default 'none'`, because breaking compatibility with significant portions of the web isn't something to take lightly -- even if they've opted in.

I do think that I understand the original concern, and the problem with having to specify an ever-growing policy of features to disable. If there were a way to distinguish between "powerful features which you want to think twice about using or granting to frames" from "fundamental features of the web platform", and apply a uniform default to those, then there might be a way to make the proposal work. Without that, it either seems very reckless on the part of the developer, or else severely impacts what kind of features can be defined (which I suspect is the more likely outcome).

Alternative strawman proposal: what if there were a way to set the default for all features which have a default allowlist of `'self'`? Those features are already disabled by default in cross-origin frames, which means that people have already at least considered the impact of having them disabled widely. The difference would be that you get to disable them all for your top-level document as well, with a single directive.

You could say something like "`default 'self'='none'`" (bikeshed colour TBD) or the much more permissive "`default 'self'=*`" without affecting the operation of features like navigation, scripting, images, forms, or such fundamental 'features'.

**Comment by valtlai** [comment]
What if we require setting up [deprecation reports](https://www.chromestatus.com/feature/5652647952515072) to use `default 'none'`.

**Comment by valtlai** [comment]
Maybe we could require setting up some report (see [Reporting API](https://w3c.github.io/reporting/)) to use `default`.

**Comment by laukstein** [comment]
@valtlai, probably best to Report things that fails because of restricted Feature-Policy header. Would basically be similar to CSP `report-uri`. I really hope we wouldn't need additional header for it.
Perhaps `report-uri` must be supported also in `Feature-Policy` header.

Since Feature-Policy are growing bigger affecting not only sensors access, but also UI like animations, fullscreen, JavaScript, etc., things gets more complicated and I hope will not break Web before spec is fixed or depreciated. For example `Feature-Policy: default 'none'` might break "all" while `Feature-Policy: default 'safe'` might be fine.

Has anybody thought how Feature-Policy would be affected by browser settings, if Feature-Policy would overwrite browser settings or opposite (I mean these might all be browser defaults or some manually customized by user)?

![grabilla g12168](https://user-images.githubusercontent.com/183966/44749635-c9a06980-ab1b-11e8-9288-d1bdc66b18df.png)

**Comment by clelland** [comment]
Feature Policy (and CSP, with the `report-to` directive) are both being integrated with the [Reporting API](https://w3c.github.io/reporting) -- see [§9. Reporting](https://wicg.github.io/feature-policy/#reporting), which makes use of an independent `Report-To` header, rather than `report-uri`.

@laukstein, I think your other question has to to with permissions -- that screenshot appears to be about setting the default permissions for various APIs. Feature Policy and Permissions do interact, as they affect many of the same features.

In general, the Permissions API is about asking the *user* whether a particular action is okay, while Feature Policy allows the *developer* to disable features, or to grant them to cross-origin frames. It's a complicated interaction, but essentially *both* of the APIs have to agree to allow a feature to be used in order to enable it. If either party (user or developer) says no, then the site cannot use the feature.

**Comment by AfroThundr3007730** [comment]
> Although that's probably not the reality of the situation -- in reality, the existence of such an option, especially if it had any significant level of adoption on the web, would mean that browser vendors would be effectively barred from introducing any new features which would break content on all of those sites. Every new feature decision would have to include serious consideration and discussion about the impact to all of the sites which use `default 'none'`, because breaking compatibility with significant portions of the web isn't something to take lightly -- even if they've opted in.

I can see this being very useful for endpoints like a REST API, which would use precisely zero of those features, and even completely static sites (e.g. readthedocs for said API) would find this useful. Just like with CSP, setting `default-src none` would break a more dynamic site immediately, but then you start allowing the origins you need (like `'self'`), enabling specific features (<s>allowing `unsafe-inline`</s>), etc. If a new feature gets added later, you go back and enable it if you need it. The same would apply here.

If part of your site suddenly became very broken overnight after a browser update, you know where to start looking (at least, that's my thought flow after having set such a restrictive security policy). For the small number of people who would find it useful to disable everything, they would likely be more aware of what probably caused breakage, should it occur. Even better if the browser console tells them what's breaking the site, like they do with CSP.

I don't see the `default 'none'` getting used by very many sites, but it should be an option for those who want to use it. The majority of sites would probably stick to the default policy or set `default 'self'` (if they implement it at all), and would be unaffected.

**Comment by sstelfox** [comment]
> I don't see the `default 'none'` getting used by very many sites, but it should be an option for those who want to use it. The majority of sites would probably stick to the default policy or set `default 'self'` (if they implement it at all), and would be unaffected.

I tend to disagree with this, especially on sites as they're developed. Setting the default wouldn't prevent you (the developer) from using any of the APIs, it just means you have to be cognizant that when you add features you are sure to whitelist them. Personally I see this as a huge benefit as a developer to prevent third party scripts and resources from doing things I don't want.

From what I've read so far about this proposal the default doesn't have to be `none` but could be any of the valid origin specifications. The two most useful IMHO are `self` and `none`.

A casual example of this (and a straw man so take it in the spirit it's given), is that of embedding a youtube video. If I have a default policy of either `self` or `none`, when embedding I could prevent the youtube iframe from attempting to enable notifications for the user (I know youtube doesn't do this, but they could in the future, as could any other site that embeds content like this).

Those requests to use functionality are outright frustrating for users when they're not expecting them and I'd much prefer to control that interaction.

**Comment by yahesh** [comment]
Please implement a default directive. I'm thinking of implementing Feature-Policies on all of my sites but am not willing to name each feature individually. So this is a setting-a-default-is-possible-and-I-adopt-this-header or setting-a-default-is-NOT-possible-and-I-DON'T-adopt-this-header decision for me.

**Comment by triblondon** [comment]
I wonder if a good compromise here is to define some 'bundle' policies. 

On the [polyfill service](https://polyfill.io) we had the problem of people needing to create a very long string of required polyfills.  The first thing we did (very unwisely) was to add an `all` alias, which includes all current and future polyfills.  This turns out to be a really bad idea (for example when we shipped `Intl`, people requesting `all` were getting every single locale pack, some 25MB of JS!).

Then we made a `default` alias, which was a movable beast that gradually encompassed more and more polyfills, but we started to find we were breaking sites by adding more polyfills into the group, even though we restricted it to the most common features only.

So what we ended up with was immutable bundle aliases, eg `default-3.6`, `default-3.7`, so every time the makeup of the 'default' bundle changed, the alias would also change.

While this comparison isn't completely perfect, I think the solution we ended up with works well for feature policy: we could have some bundle names like `antipatterns-1`, `user-consent-1`, and if a new policy is added that guards an antipattern, then a new antipatterns policy is also created.

**Comment by ExE-Boss** [comment]
I think plain `default 'none'` should disallow everything that isn’t pure JavaScript (i.e.&nbsp;no&nbsp;DOM, no&nbsp;`window` global (`globalThis` would be&nbsp;used instead), no&nbsp;nothing).

After that, you would enable features using other directives.

**Comment by Jamesernator** [comment]
I would propose an alternative of requiring a date to be provided and that whenever a feature is implemented it has an associated date.

For example we might have the following policies with their dates of addition:

```txt
foo-bar : 2018/07/27
baz-bar : 2018/01/23
boz-bar : 2020/11/14
// etc
```

Now if an header like so is given by a site: `FeaturePolicy: default 2019/01/01 'none'` is given then `foo-bar`/`baz-bar` would be included but `boz-bar` would not.

---

How dates would be associated with a policy wouldn't really matter, all that would really need to be a requirement is that a browser couldn't add a new policy for a date in the past.

**Comment by Salamander3** [comment]
Who put this stubborn person in charge of this again? @clelland You just won't accept feedback, the most narrow-minded person i've seen.

**Comment by ojanvafai** [comment]
Please keep discussion focused on productive debate of the issues. I've deleted the most recent comment on this issue as it consisted solely of a personal attack, which violates the W3C code of conduct (https://www.w3.org/Consortium/cepc/). Repeated violations will result in being blocked from commenting on the repository.

**Comment by sstelfox** [comment]
Reading through some of the more recent responses it seems like there are two independent groups talking about different meanings of how `default` should behave.

Originally it was about setting the policy for any features not explicitly otherwise listed in the rest of the feature policy. Specifically, "If my feature policy doesn't mention feature X what policy should be applied". This covers new features that haven't been defined when the policy was configured and prevents having to specify `none` or `self` on each of the many different features that most sites won't use.

The second camp that has recently started up, seems to be more discussion what the default setting for each feature should be and have it defined by the w3c policy rather than a developer set default. This is along the lines of "use the default feature settings as defined by the w3c as of 2019-02-28". As other have mentioned, this would be a thorny and contentious road. There is likely no good default set that won't hurt sites out there.

There likely shouldn't be defaults when the header isn't present as has been done with every other security control based on HTTP headers, and when the header is present it only makes sense to allow the developer to set the default policy for everything, then override specifics.

Back to the original feature, the only argument against being able to set a default policy (please correct me if I've missed something) seems to be that a future feature addition might break existing sites that haven't changed. I don't think this will ever be an issue as long as an existing feature is never broken up into more granular features.

For example if `Clipboard` got broken up into `Clipboard-Write` and `Clipboard-Read`, and `Clipboard` was deprecated the default setting might break existing web apps that rely on doing one of those functions. Preventing this kind of split can be done at the w3c policy level and as long as the developers of the policy are cognizant of this risk there shouldn't be any issue.

In any other case, the only time a default setting would break any functionality is when new functionality is being added to the web app that requires a change to the feature policy. In this scenario the web app is actively being changed and the developer will already have to deploy a new version of the code (it's also worth noting that this would be adding new features to the site, not just a minor bug fix release).

In this case part of the release of the new version would require the header being updated as well to allow for the new feature. This is the same level of effort as other security mitigations such as CSP headers and I think is completely reasonable.

**Comment by Jamesernator** [comment]
> This is along the lines of "use the default feature settings as defined by the w3c as of 2019-02-28". As other have mentioned, this would be a thorny and contentious road. There is likely no good default set that won't hurt sites out there.

Actually my proposed idea doesn't depend on the W3C managing the dates just that a browser never adds a new policy for a date in the past, each browser could manage their own list simply based on commit date or similar.

The date would be a "my site works at this date" hence if a new policy is added it should not affect any page which gives a date before the date of the new policy, this doesn't require browsers to agree or even have any rules around it other than "a date for a new policy must not be in the past". This is why I think the date strategy is feasible and very non-thorny or should even be contentious.

**Comment by clelland** [comment]
The work going on around issues #296, #298, #299, #300, etc, should help clear up a lot of the issues around this. If the proposed document policies and sandboxing are not covered by the Feature-Policy header, then a default allowlist makes a lot more sense; whether it is 'self', or 'none', or some concrete list of trusted origins.

**Comment by sstelfox** [comment]
I'm pinging this again because it doesn't seem like there has been any movement on this and it seems both a very easy change and a seriously impactful one. Default deny (or in this case the ability to at least specify default deny) has been the best practice for security features for a very long time.

This already exists for the content security policy header so there is a very relevant reference. It doesn't break anything as an administrator or developer has to actively add in the default header and it can't break a working application unless changes are already being made to this.

It's crazy to me that this is still an open proposal as there is no downside to specifying it.

**Comment by oliverjanik** [comment]
Just setting up new site and looking at this again. Are there any news? 

Ideally I'd like:
1) deny all
2) allow only my specfic set

The amount of directives has grown quite a lot: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy

It's not feasible to enumerate all of them.

**Comment by Valodim** [comment]
I have also stumbled upon this issue, while attempting to disable "all" features on a page with plain content. Enumeration of all undesired features scales badly in terms of future compatibility and size. I decided against adding feature-policy headers for the time being: the available policies seemed overly specific, yet at the same time not broad enough to match my intent. It seems clear to me that a blacklist isn't the way to go.

However, I can see why browser vendors are reluctant to include a blanket "disallow all" feature policy. It would make the process of introducing additional features in the future much more complicated, particularly those that would allow sandboxing functionality that currently feels "fundamental", like displaying images, or influencing navigation.

I don't buy the argument that this is what admins buy into when they disallow all features. "Best practice" checkers already tell you to set feature-policy headers, and not everyone keeps up to date with standards development. People *will* disable all features if it works for the moment, and then pages *will* break because noone can realistically monitor that their pages work in all browsers forever. And that breakage will come back to browsers.

Neither of the extreme options seem viable to me. Two middle ground solutions were already brought up above: introducing "bundles" to allow reasoning about groups of features (e.g. "sensors", "media", "hardware"), or timestamp features to give intended behavior a temporal context. I would encourage everyone to engage with those, or other options in this direction.

**Comment by kovge** [comment]
My opinion is, that the feature policy clould follow the concept of content policy in case of a default value.
We start to use the policy from scratch (default 'none/off/...')
If we do not implement this header,  we let the browser / user to allow / disallow ... as before this policy was created.

If a developer works with a feature, he/she can than add the policy feature to envirolment setup (server side headers via vhost,htaccess,server side script and so on) And write an install.md sectuon with used features. It makes anyway more transparent and controlled (without reading maybe thousands of code lines of a project) any third party client side code.
Think about a schenario. I start to create a web app. If i have a chance to turn off all the features, i can integrate a 3rd party library, without having the risk, that the library creator forget to block a policy and works anyway.
Second schenario: you let your customers to load 3rdbparty javascript from trusted domains.
You blocked all unneccesarry features, than tomorrow appears a new one, than they implement it, and without an audit it does his thing.

In most cases we start from a base, than add more and more features to our web application. Sonlogically we start from " no feature is needed ti work with it ". 

My opinion is, that javascript / css .. are also "features" on their own way, hust since web2.. and so on, they became as a part of standard web envirolment (including the security risks they hold, thats why we need csp headers).

"Features" are different things. Accessing to device resources. But regarding privacy and so on,  they do not hold less security risk. (F.e: camera turns on in near of every atm. I know: you need a running program code thatfor, but a regular user has no chance to analyze the code)
So it could give more trust (and more help for code audit) if we say "i do not let my web app to access any features". It helps for team work on web applications. We start from a 0 state, and as we integrate or implement new features, we just turn them on, and include to privacy polici and whatever the Law says...

It is like a firewall on its own way. As most of them starts from: All are blocked, and we open ports/services if we know what we do.

So one more big like for this feature request. How will be defined? It is not as important. But To have something to turn all features off, is important. It is security feature.
If someone fears that he/she bricks his web app, than he wont use it. In other hand, that web app will be less trusted ( maybe in feature blocked by browserd).

And one hint: just keep the concept, witch works perfect for content policy. (Maybe we do not need here domain and unsafe inline and so on, just a on/off/may ... ) as we have there default... why not here as well. Better to follow one concept regarding (client side) web app security.

**Comment by ConstableBrew** [comment]
How is this still not implemented? A simple REST API specifying 'none' for all features would add a ridiculously unwieldy 699 bytes to the header *for every request*!

**Comment by gertvdijk** [comment]
@ConstableBrew 
> How is this still not implemented? A simple REST API specifying 'none' for all features would add a ridiculously unwieldy 699 bytes to the header for every request!

I was about to comment this too. However two thoughts to consider too:

- Use of HTTP/2+ with HPACK will reduce the bandwidth on headers sent repeatedly on a connection (but yes, it is still bytes coming out of the HTTP library used over and over).
- IIUC, you don't have to use this header on API request. Just the HTTP response for the landing page (HTML) loaded in the browser should be sufficient to apply the feature policies for the lifetime of the page/DOM, no?

**Comment by clelland** [comment]
FWIW, I agree that we should do this. One of the motivations behind Document Policy was to allow Feature Policy to focus on the 'powerful features' use case -- to have it mostly be about handling those features which you would generally not want to have enabled in third-party contexts. Having a mode which says "Start with the settings of a third-party iframe, and then enable the following things" seems reasonable.

The biggest factors in the way of this have been:
  - Syntax, especially with parameterized features
  - Overreach of feature policy to handle not just powerful features, but performance best-practices and other document configuration
  - Not wanting to put developers in a place where a default:none directive is extremely like to break their site in the future
  - Conversely, not wanting to put browser vendors in a place where widespread deployment of default:none means that no features can be added in practice

With Document Policy available to handle the other sorts of features, dropping the proposal for parameterized features from feature policy, and the possibility of a new header format (from #359), this looks within reach.

**Comment by xdevs23** [comment]
Considering that I'm currently using
```
ambient-light-sensor 'none'; autoplay 'none'; accelerometer 'none'; camera 'none'; display-capture 'none'; document-domain 'none'; encrypted-media 'none'; fullscreen 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; midi 'none'; payment 'none'; picture-in-picture 'none'; speaker 'none'; sync-xhr 'none'; usb 'none'; wake-lock 'none'; vr 'none'; xr-spatial-tracking 'none'
```

as policy, I think it does make sense to simply specify

`default 'none'`

**Comment by laukstein** [comment]
In recent years we are adding more and more headers for privacy and security, and its size is rapidly increasing. While Internet infrastructure is evolving, we must not ignore low RAM devices, performance, speed and low bandwidth. I am thinking where secured IoT device could be in size of 1 millimeter ... and I am worried with current spec having headers size larger than the response content itself... Imagine 80kb headers size for every second API call with response of 0.01kb content ...

It likely would take many many years till having a spec like automatically fetching and caching package.json for domain root (can compare concept of manifest.webmanifest and robots.txt) that specifies rules of "all headers" having defaut state and per url regex. I recommend to take it already in spec consideration-discussions before "breaking Web by spec".

**Comment by xdevs23** [comment]
> Imagine 80kb headers size for every second API call with response of 0.01kb content ...

this does not really apply, you don't need those headers when you're calling an API from a microcontroller since these headers will only be used by browsers.

**Comment by laukstein** [comment]
> this does not really apply, you don't need those headers when you're calling an API from a microcontroller since these headers will only be used by browsers.

I am not referring only for this specific header but for headers globally.

**Comment by xdevs23** [comment]
> I am not referring only for this specific header but for headers globally.

Yeah, still, only browser use these headers so you can just remove them if the API is only for microcontrollers.

**Comment by vphantom** [comment]
We're in a situation where the same web service answers web page requests and `XmlHttpRequest` exchanges, and our security scanner warns us if this header is missing.  As a result, we're including the litany of possible `none` (obtained by combining the documented values from several browsers, probably still incomplete) in all our responses.  This is wasteful but unavoidable since our only alternatives would be to remove this header and tolerate the security warnings, or perhaps to reduce the header to just the most obvious like `geolocation`, `camera` and `microphone` to keep it shorter.  Neither of those alternatives would give clients accurate information.

If the header let us declare what we intend to use, instead of forcing us to list everything we are aware of that we promise not to use, this wouldn't be an issue and it would be future-friendly as well, always remaining accurate when new values are created.

**Comment by xdevs23** [comment]
> We're in a situation where the same web service answers web page requests and XmlHttpRequest exchanges, (...)

You should be able to specify headers depending on what you're requesting in your web server configuration. But yes, I still second the OP.

**Comment by ewanm89** [comment]
> 
> 
> > I am not referring only for this specific header but for headers globally.
> 
> Yeah, still, only browser use these headers so you can just remove them if the API is only for microcontrollers.

You are assuming (and every one know the result of when you assume something), that browser will never consume that service at any time. Which is a pretty big assumption (I consider having an API that device sends data too, and then user logs into via a web interface for later display and processing of data).

**Comment by xdevs23** [comment]
> You are assuming (and every one know the result of when you assume something), that browser will never consume that service at any time. Which is a pretty big assumption (I consider having an API that device sends data too, and then user logs into via a web interface for later display and processing of data).

Point still stands, APIs don't need all of those headers. You don't need to specify Feature-Policy, Content-Security-Policy etc. because that only works with HTML, not APIs. What you should specify are CORS and HTTPS-related headers.

**Comment by oliverjanik** [comment]
While you're technically correct @xdevs23 the pentesting companies apply blanket scanners that flag missing headers and then non-technical management gets reports of security "vulnerabilities" that become top priority and you end up back here in this thread.

**Comment by xdevs23** [comment]
> While you're technically correct @xdevs23 the pentesting companies apply blanket scanners that flag missing headers and then non-technical management gets reports of security "vulnerabilities" that become top priority and you end up back here in this thread.

That's a fair point. But in that case there is no reason to worry about header size in general. Only Feature-Policy is one that requires maintenance (features keep getting added or removed).

**Comment by ewanm89** [comment]
It is entirely possible for a) put the URL in a browser for a Web API even if a computer is more likely to consume it, and b) for the api to output data in a html based format for other use. There are reasons pentesting scanners always flag on this header, if these headers are to have any real protection, they must **always** be used, that way you don't miss some odd use case cause you aren't using it that way.

Hell has everyone at W3C forgotten that XHTML, XML and XSLT exist?

**Comment by dimaqq** [comment]
Here's my attempt for a battened down admin site: `accelerometer 'none'; ambient-light-sensor 'none'; autoplay 'none'; battery 'none'; camera 'none'; display-capture 'none'; document-domain 'none'; encrypted-media 'none'; fullscreen 'none'; geolocation 'none'; gyroscope 'none'; legacy-image-formats 'none'; magnetometer 'none'; microphone 'none'; midi 'none'; oversized-images 'none'; payment 'none'; picture-in-picture 'none'; publickey-credentials 'none'; sync-xhr 'none'; unoptimized-images 'none'; unsized-media 'none'; usb 'none'; vibrate 'none'; vr 'none'; wake-lock 'none'; xr-spatial-tracking 'none';`

What's wrong with that:
* adds 558 bytes (uncompressed) 259 bytes (gzip) or 226 bytes (br) to **first** request to the site
* too long for supported by some (👉Azure CDN) servers
* insane to maintain

**Comment by Malvoz** [comment]
If y'all can take a moment to read [this](https://github.com/w3c/webappsec-feature-policy/issues/189#issuecomment-513320403) and [this](https://github.com/w3c/webappsec-feature-policy/issues/189#issuecomment-621877304) you'll quickly realize `Feature-Policy` is still in an experimental/evolving state. The way people have been asking to "just do this" in the current state of FP means if a feature such as [disabling javascript](https://github.com/w3c/webappsec-feature-policy/issues/365) comes around, you're okay with having scripts disabled too?


Off-topic: @dimaqq FYI [`vibrate`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/vibrate) isn't defined in any spec, nor implemented in any browser so you could remove that, `publickey-credentials` was renamed [`publickey-credentials-get`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/publickey-credentials-get), and [`vr`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/vr) is deprecated as the WebVR API was replced with the WebXR Device API ("`xr-spatial-tracking`").

**Comment by sam-spain** [comment]
Is there a more maintainable way to do the following for Apache server?
`Header always set Feature-Policy "ambient-light-sensor 'none'; autoplay 'none'; accelerometer 'none'; camera 'none'; display-capture 'none'; document-domain 'none'; encrypted-media 'none'; fullscreen 'none'; gyroscope 'none'; magnetometer 'none'; microphone 'none'; midi 'none'; payment 'none'; picture-in-picture 'none'; sync-xhr 'none'; usb 'none'; wake-lock 'none'; xr-spatial-tracking 'none'"`

It seems really silly to have this. Also means any new features will need to be added.

**Comment by laukstein** [comment]
Doe to policies list repeatedly growing, in order to handle it reasonable, the policies should be grouped in "categories" and maybe also "sub-categories" allowing to use like:

    Permissions Policy sensors 'none'; microphone 'self'; dom-changes 'none'; fullscreen 'https://www.youtube.com'

Groups and sub-groups: e.g. sensors (media), JavaScript (DOM manipulations, workers), etc.

**Comment by clelland** [comment]
Groups are something that's been considered, that hasn't been specced yet -- mostly because I haven't gotten around to it, but it's certainly becoming a bigger issue. Adding client hints to the set will expand this list quite a bit. It would certainly be more ergonomic to have a way to group features together.

I suspect that there are probably two different things we could do there: One is to provide less-granular control, like in https://github.com/w3c/webappsec-permissions-policy/issues/189#issuecomment-694408105, to configure features in well-defined groups, where there might even be a top-level group for all features (but see below)

The other, that I mentioned in https://github.com/w3c/webappsec-permissions-policy/issues/189#issuecomment-416308667, would be to provide a single setting that essentially says "treat this as an untrusted document, as if it were loaded in a cross-origin frame". That setting would ensure that the basic platform features continue to work, but that none of the more powerful features are available. Anything which is restricted by default in cross-origin frames would be blocked.

I like the idea of groups, but they would need to be well-defined, and there would have to be a way to ensure that there was no overlap (although nested groups might be possible). We'd need to be very deliberate about which group each feature belonged in, to avoid any ambiguities if one group is disabled while another is enabled.

**Comment by bubsly** [comment]
Whilst I agree the whitelist approach makes the most sense, it would be a www breaking change.
I would propose that a new header be implemented that is a whitelist version of the previous (Permitted-Permissions-Policy?)
All new/upcoming policies would be disabled by default and would need to be defined under this new policy. If the policy is present, ignore the older 'Feature-Policy/Permissions Policy'. I would also reverse the order so that * and 'self' came first.

This would allow for older systems to not break and would also not subject them to the security concerns of newer features.
As for the directives:

**Wildcard**
`*`
> e.g. all=*


**Groups:**
`sensor: [ accelerometer, ambient-light-sensor, gyroscope, magnetometer ]`
`capture: [ camera, geolocation, usb, vibrate, microphone, vr, display-capture, midi]`
`system: [wake-lock, battery]`
`html: [xr-spatial-tracking, autoplay, fullscreen, encrypted-media, legacy-image-formats, oversized-images, payment, picture-in-picture, sync-xhr, unoptimized-images, unsized-media  ]`
`authentication: [publickey-credentials]`
`legacy: [document-domain] `

> group selector. e.g. self = [sensor], [html]; all = [authentication]; example.com = [system];

**Individual**

> e.g. self = battery

A complete policy would look like this:
`Permitted-Permissions-Policy: all = [html]; self = geolocation, [sensor]; google.com = geolocation, payment, [authentication];`

**Comment by winter2x** [comment]
If a conscious decision is made by an entity to use a whitelist approach, one should infer from that their willingness to manage the policy so that their site remains functional.
Removing this ability weakens the security aspects of the header making it a counterproductive decision.

So my proposal:
By default the rule should always prepend `*=default`, subsequent policy declarations override that, e.g.: `*=()`.
Wildcards could also provide group-like functionality, so sensor-accelerometer, sensor-ambient-light and other sensor functionalities could be defined as `sensor-*`.
If I wanted to make the header as small as possible I could think of something like `*=(); (sensor-* media-autoplay)=self; sensor-gyroscope+=("https://mysite.com" "https://othersite.com"); sensor-ambient-light=()`.

**Comment by vphantom** [comment]
Quick question: why would such a feature whitelist come with an entirely new syntax to parse and produce where the general format of `Content-Security-Policy` already exists for a very similar purpose?  For example, perhaps:

```text
Permissions-Policy: default 'none'; geolocation 'self'; media-autoplay some.othersite.com 'self'
```

**Comment by winter2x** [comment]
Quoting the explainer document:

----

A policy which would previously have been expressed as:

```http
Feature-Policy: fullscreen 'self' https://example.com https://another.example.com;
    geolocation *; camera 'none'
```

would now look like:

```http
Permissions-Policy: fullscreen=(self "https://example.com" "https://another.example.com"),
    geolocation=*, camera=()
```
----

For my proposal I took the new format and made minor alterations to its syntax to shorten the length of the header while making it easier to maintain and understand at a glance, at the same time also offering a solution for implementing feature groups.

In the current format the policy I described, provided setting *all* to *none* was a possibility
`*=(); (sensor-* media-autoplay)=self; sensor-gyroscope+=("https://mysite.com" "https://othersite.com"); sensor-ambient-light=()`
would look like this:
`*=(), media-autoplay=self, sensor-gyroscope=(self "https://mysite.com" "https://othersite.com"), sensor-sensor-1=self, sensor-sensor-2=self, ..., sensor-sensor-n=self`

**Comment by gapple** [comment]
The header format is [Structured Field Values for HTTP](https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html) (which didn't exist when CSP was defined); the header is a dictionary, keyed by feature name and each dictionary value an inner list of origins.

`(sensor-* media-autoplay)=self` would not work with the Structured Field Values syntax, and would need to be specified as `sensor-*=(self); media-autoplay=(self)`

**Comment by winter2x** [comment]
Yes, unless there is a change to allow inner lists as keys, for which there could be a need now.

**Comment by runthis** [comment]
I know this discussion is leaning towards 3 years old now. Does it seem possible in 2021 to disable all by default? On my portfolio website I don't foresee needing any permissions and would happily enable them one by one if I did (much like how CSP works).

**Comment by sandorvasas** [comment]
I've just come across this header and the intuitive way to me would be to disable all by default and explicitly enable stuff in the header. I'm surprised this is an unresolved issue.

**Comment by olmari** [comment]
Mine 2 cents is still basically as @sandorvasas said, in additition with possibility in header to define default that caters to everything. Thus if nothing is done, nothing happends, if site admin has defined default then that is used if no direct match exist.

**Comment by Seirdy** [comment]
Adapting something I posted on Fedi a few days ago:

Every new web feature wants its own entry in this massive chonker of a header, and there’s no one place to view all the directives. Here’s where we’re at now:

```
Permissions-Policy: accelerometer=(),ambient-light-sensor=(),attribution-reporting=(),autoplay=(),battery=(),camera=(),clipboard-read=(),clipboard-write=(),conversion-measurement=(),cross-origin-isolated=(),direct-sockets=(),display-capture=(),document-domain=(),encrypted-media=(),execution-while-not-rendered=(),execution-while-out-of-viewport=(),focus-without-user-activation=(),fullscreen=(),gamepad=(),geolocation=(),gyroscope=(),hid=(),idle-detection=(),interest-cohort=(),magnetometer=(),microphone=(),midi=(),navigation-override=(),otp-credentials=(),payment=(),picture-in-picture=(),publickey-credentials-get=(),screen-wake-lock=(),serial=(),shared-autofill=(),speaker-selection=(),storage-access-api=(),sync-script=(),sync-xhr=(),trust-token-redemption=(),usb=(),vertical-scroll=(),wake-lock=(),web-share=(),window-placement=(),xr-spatial-tracking=()
```

That one-line view doesn't quite capture the current state of things. Here's a multiline list of current directives:

```
     1	accelerometer
     2	ambient-light-sensor
     3	attribution-reporting
     4	autoplay
     5	battery
     6	camera
     7	clipboard-read
     8	clipboard-write
     9	conversion-measurement
    10	cross-origin-isolated
    11	direct-sockets
    12	display-capture
    13	document-domain
    14	encrypted-media
    15	execution-while-not-rendered
    16	execution-while-out-of-viewport
    17	focus-without-user-activation
    18	fullscreen
    19	gamepad
    20	geolocation
    21	gyroscope
    22	hid
    23	idle-detection
    24	interest-cohort
    25	magnetometer
    26	microphone
    27	midi
    28	navigation-override
    29	otp-credentials
    30	payment
    31	picture-in-picture
    32	publickey-credentials-get
    33	screen-wake-lock
    34	serial
    35	shared-autofill
    36	speaker-selection
    37	storage-access-api
    38	sync-script
    39	sync-xhr
    40	trust-token-redemption
    41	usb
    42	vertical-scroll
    43	wake-lock
    44	web-share
    45	window-placement
    46	xr-spatial-tracking
```

The above example excludes client-hints permissions which allow delegating permission to read client hints to other domains.

You can view all the perms supported by the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/tot/Page/#type-PermissionsPolicyFeature) at once. A subset is documented in [a non-normative companion document](https://github.com/w3c/webappsec-permissions-policy/blob/main/features.md) to the main Permissions-Policy spec.

I found some scattered documentation for a few perms that have been implemented by Chromium:

- [otp-credentials](https://web.dev/web-otp/)
- [direct-sockets](https://github.com/WICG/raw-sockets/blob/main/docs/explainer.md)
- [storage-access-api](https://privacycg.github.io/storage-access/#permissions-policy-integration)

Pages that don't use any sensitive APIs should be able to just opt-out of this permissions just like what `sandbox` allows. A `default` directive would allow  these pages to switch to selective-allowing while still letting others ignore it and perform blocking on a case-by-case basis.

HTTP headers are already a significant chunk of total page weight, even with HPACK. This header is easily the biggest on simple pages.

**Comment by dimaqq** [comment]
### Off-topic

I feel that the feature creep will kill this entire proposal (not the defaults idea).
Everyone and their uncle can add yet another directive (floc, otp) which becomes otp-out for existing web sites as it is now.
(opt-in if the defaults idea becomes the norm)

My counter-proposal would be: majority (all?) of these features require client-side JavaScript, why not set up these restrictions using JavaScript as well?
Case in point would be `requestStorageAccess` for `storage-access-api`.

**Comment by theherk** [comment]
It still seems like, even if you use several groups, it should be the option of the site's administrator to choose explicitly to allow or disallow _all_ features, even if that means a breakage when permissions are added. As long as it isn't the default, this makes sense.

**Comment by jsleeio** [comment]
I arrived at this issue when, like some other folks, I just wanted to disable it all, and disable all future new features also, with the knowledge that I'll need to enable them if I want them. The situation that has me concerned is the intersection of three possibilities, none of which are unusual or unlikely:

1. new capabilities being added to browsers that I definitely don't want to be used on my site
2. third-party-hosted Javascript things that I am intentionally using, such as in-browser performance monitoring tools, comment platforms, etc
3. changes — whether legitimate or via a security breach — to the Javascript in (2) to make use of those new capabilities

Whitelisting means there should be no window of time in which those new capabilities are used on my site. I don't know how else to prevent that.

The current approach appears to require responsible site operators to keep up to date with **_all_** new browser tech rather than focusing on the tech they're actually using. I don't think that's very respectful of their time

**Comment by Seirdy** [comment]
I thought of a solution that should address concerns surrounding both header bloat and future proofing:

Bundle all current permissions-policy directives into a single new directive called `2022`. A site with `Permissions-Policy: 2022=()` would disable all the aforementioned permissions.

In 2024, introduce the `2024` permission which includes all of `2022` as well as any permissions introduced since then.

Agents seeing a number in the permissions-policy header for a year they don't know of (e.g. `2384`) can just pick a lower year that they support.

Rinse and repeat.

**Comment by olmari** [comment]
Why is this is hard? Nothing else will work well, but treat without any policy as usual, and have policy to have option for "default" which caters for any current and future things (be it block all or whatever "default" is considered at any point in time, doesn't really matter in context), and then the fine-grained control for those who want to do it properly... and even then leave possibility to have _rest_ of stuff as default so future is catered... I can't believe this is gone on for years on end while only sane solution is simple...

**Comment by vphantom** [comment]
This header was quite simply designed wrong and needs a backwards-compatible syntax added to have its use reversed. CSP was done right and I have yet to understand why this one wasn't. Being able to disable all and to enable specific permissions is the only sustainable way to go.

**Comment by kifd** [comment]
Yet another +1 to wanting a default policy setting, just like CSP. How is it considered best practice to individually blacklist an ever growing feature set, rather than whitelist the features you actually want?

**Comment by vphantom** [comment]
Is there any other way to get the W3C's attention about this?

It's been over 5 years and we've been through one header change already (FP → PP) so clearly they either don't care or never even noticed this major design flaw which is costing everyone (interested in compliance) time to maintain the list of features, and bandwidth to send those headers to browsers.

A simple `all=()` ahead of the more specific permissions we _do_ use (if any) would be fine and backwards-compatible.

**Comment by dimaqq** [comment]
> Is there any other way to get the W3C's attention about this?

☝🏻 easily, don't use this proposal, let it die.
If the origin servers don't send this header field, the browsers will not enforce it.
Someone will eventually come up with a better proposal, or a better concept.

**Comment by theherk** [comment]
@dimaqq I think I get your point, but I don't think just throwing our hands up will do much either, as we still have to make these things work and abandoning CSP simply isn't going to happen. If browsers stopped supporting it, we'd be going the wrong direction. Better the proposal is ideally adopted at some point. If a better proposal exists or comes up, then that proposal can be adopted. In the meantime, this seems to be fairly well-liked.

**Comment by Gunni** [comment]
I don't think anyone here is talking about the content security policy since it has a sane default policy. The permission policy on the other hand does not which is what we are talking about here.

**Comment by theherk** [comment]
Yes, I misspoke. But I still maintain that the abandoning the proposal without an alternative doesn't seem better than this proposal.

**Comment by vphantom** [comment]
In our case we're forced to output that header in order to pass basic security scans so I can't let it die. :( At best I could reduce the header to almost nothing, but that would defeat the purpose of promising to behave. As it stands, our header has 62 of the 66 known features explicitly disabled and I have to search for new features regularly. I could probably remove some benign stuff to bring it down to 20-30 but I'd still have to regularly hunt for new offenders.

There have been several well-designed, backwards-compatible suggestions here these past 6 years and yet not a peep from the W3C to at least acknowledge the issue. They even mention in <https://www.w3.org/TR/permissions-policy/> that "The web platform provides an ever-expanding set of features" and yet still maintain that selectively disabling individual features we know about is appropriate.

It is somehow the webmaster's responsibility to discover new features as they arise and to add them to an ever-growing deny list in order to correctly promise good behavior and to help deny access to misbehaving third parties (included JS, PunchOut frame parents, etc.)

**Comment by yannikbloscheck** [comment]
The header obviously gets way too big, if you actually want to disable most things. It's just frustrating. This really needs to be finally solved. 
It's obvious a default directive is needed. No matter if it actually will be called `default=()`, `all=()` or anything else.
I really can't understand why this is being ignored for so long. Apparently this has been discussed and requested for over five years now...

**Comment by Seirdy** [comment]
> The header obviously gets way too big, if you actually want to disable most things. It's just frustrating. This really needs to be finally solved. It's obvious a default directive is needed. No matter if it actually will be called `default=()`, `all=()` or anything else. I really can't understand why this is being ignored for so long.

@yannikbloscheck If you read the discussions you cited, you would know that a plain default-deny directive would not be forwards-compatible and is therefore not an option. Defining new meta-directives with specific static inheritance rules is a better option.

**Comment by yannikbloscheck** [comment]
> @yannikbloscheck If you read the discussions you cited, you would know that a plain default-deny directive would not be forwards-compatible and is therefore not an option. Defining new meta-directives with specific static inheritance rules is a better option.

@Seirdy I read your ideas.
The directive groupings for feature areas is a good idea, but practically those group directives would need to be extendable, which would lead to the same problems. I still hope such groups will be implemented though, because it's a great idea anyway.
The directive groupings by point in time (or probably rather version number) would solve many of these problems. Web developers still would need to update the policy regularly to deny as much as possible. Also browser developers would need to agree every time what is included in each version. Also not a bad idea either.

A default directive would simply leave the choice to the developer though.
Maybe I don't care about being forwards compatible with any new (or even existing) access. Then I choose `default=()`. Maybe I want to allow anything new for my site by default. Then I could choose `default=(self)` etc.
If I want to be a responsible developer, I have to keep myself updated with what is added to this header in all cases anyway.
There are many other security headers that a web developer can use to shoot their future self in the foot by using too restrictive choices. But all those headers still exist anyway.

I can still understand the fears the browser developers have about this in regards to forward-compatibility. Especially as they don't really have to worry about all the extra unnecessary traffic the current solution causes. The browser developers probably would be forced to heavily communicate some new directives and have longer roll-out times. Still any developer worried about having anything break ever could just leave the default directive out or set a very permissive one. The documentation by the browser developers etc. could for example suggest `default=(self)` as the best choice.

The main thing though is that this has been a problem for over five years. While there has been much discussion about whatever problems might happen with any solution in the future, there has been a real problem (incredible header bloat) in practice for all this time and it isn't getting solved.

**Comment by PeteX** [comment]
It's great that this is moving forward, thanks for coming up with this proposal.

One question—will browser developers ever want to change the set of permissions that are applied to third-party iframes?  Recently they have become more locked down, to reduce cross-site tracking of users.  If that trend continues and there is a wish to withdraw some of the permissions we're discussing here, it could lead to divergence between the "untrusted" permissions policy and the permissions that are granted to third-party iframes in practice.

I'm genuinely not sure if this is an issue or not, so I will now leave this for people more qualified than I am. 🙂

**Comment by annevk** [comment]
The concern @PeteX raises is one I share and I continue to believe we should not do this therefore. Although it is somewhat painful, it is possible to enumerate all the permissions and disable them. That in turn allows us to make further tweaks to what is allowed in nested documents more easily, such as possibly gating the `<input type=file>` dialog behind a permission.

**Comment by jvoisin** [comment]
> The concern @PeteX raises is one I share and I continue to believe we should not do this therefore. Although it is somewhat painful, it is possible to enumerate all the permissions and disable them. That in turn allows us to make further tweaks to what is allowed in nested documents more easily, such as possibly gating the <input type=file> dialog behind a permission.

Is it though? I don't know all the permissions on top of my head, and I shouldn't be required to to simply configure my website as a *regular* one. Moreover, even if I somehow found a somehow comprehensive list like `allow="accelerometer 'none'; ambient-light-sensor 'none'; autoplay 'none'; battery 'none'; bluetooth 'none'; browsing-topics 'none'; camera 'none'; ch-ua 'none'; display-capture 'none'; domain-agent 'none'; document-domain 'none'; encrypted-media 'none'; execution-while-not-rendered 'none'; execution-while-out-of-viewport 'none'; gamepad 'none'; geolocation 'none'; gyroscope 'none'; hid 'none'; identity-credentials-get 'none'; idle-detection 'none'; keyboard-map 'none'; local-fonts 'none'; magnetometer 'none'; microphone 'none'; midi 'none'; navigation-override 'none'; otp-credentials 'none'; payment 'none'; picture-in-picture 'none'; publickey-credentials-create 'none'; publickey-credentials-get 'none'; screen-wake-lock 'none'; serial 'none'; speaker-selection 'none'; sync-xhr 'none'; usb 'none'; web-share 'none'; window-management 'none'; xr-spatial-tracking 'none'"`, maybe it'll be outdated in 2 months when someone adds yet another weird permission, who knows. Also, maybe I lied and the list isn't complete, can you find out what's left out? CSP is a security control, it should be straightforward to audit: the current "spell everything out and hope you don't forget anything" state isn't.

**Comment by annevk** [comment]
> maybe it'll be outdated in 2 months when someone adds yet another weird permission

But that is the point. Your website might be making use of a feature that the weird permission would gate. If you opted into everything your website would break and end users would be sad. This is a risk we have been willing to take at times for third-party content, but this would extend the risk to every website, making it far less tenable to do.

**Comment by jvoisin** [comment]
I'd argue that if you want to make use of the latest new feature, it's pretty easy to add `; new-shiny-feature 'self'`, versus putting the burden on every single other website that doesn't use it. Moreoever, even if the majority of websites would like to use this feature, they'll still fail-open for the next added one, which is a terrible idea security-wise.

**Comment by yahesh** [comment]
It's really saddening to see that even half a decade later the discussion still evolves around the same thing.

As long as this fails open when set, it's not a security header (even though different security scanners started to handle it as such). As long as there's no deny-all possibility, I'll continue to tell everyone off who expects this header to be implemented as a form of security improvement.

**Comment by ewanm89** [comment]
> The concern @PeteX raises is one I share and I continue to believe we should not do this therefore. Although it is somewhat painful, it is possible to enumerate all the permissions and disable them. That in turn allows us to make further tweaks to what is allowed in nested documents more easily, such as possibly gating the `<input type=file>` dialog behind a permission.

This fails security 101 of The principle of least privilege I therefore hand you a fail in security 101, the idea here is to deny the vast amount of permissions by default thereby gating most of them you can still enable them.

I do think they should be a deny all and not just some.

The principle of least privilege is not a nice thing, it is how you build secure systems, and developers can always enable to features they use and need. We already need to spin up a local webserver to get a secure context for testing, we are already doing this work. We should have a way to default to none so we can't miss one by mistake, or because browser A implemented something I don't know about because browser B is my prefered browser. If there is a problem I can enable the permission.

**Comment by darioseidl** [comment]
> > maybe it'll be outdated in 2 months when someone adds yet another weird permission
> 
> But that is the point. Your website might be making use of a feature that the weird permission would gate. If you opted into everything your website would break and end users would be sad. This is a risk we have been willing to take at times for third-party content, but this would extend the risk to every website, making it far less tenable to do.

This point has been made, and countless people have voiced their opinion already that it does not apply to them: I want to author a website that is secure and works now. I want it to be secure in the future, but I don't mind if it breaks in the future.

If browsers introduce a permission policy for an existing feature that my website is using already, one of the following things will happen:

1) I will read about it before it is introduced, and I can decide whether I need that feature or not. It's all good. As developers of any software, we already have to stay informed to keep things working, given that software changes constantly. It's not the end of the world.

2) I miss the news about the policy and my website breaks (or rather, I would expect some specific part of the website to stop working). My users are sad. But I test my website too, and notice that something has stopped working. Or my sad users tell me, please fix this. So I fix it. It's not the end of the world.

If somebody doesn't want that to happen to them, they simply don't need to use the proposed deny-all policy. 

Please just introduce a deny-all permissions policy. It's the sane thing to do.

**Comment by PeteX** [comment]
I'm sorry! 😢 I posted what was meant to be a question about a minor detail, then went away for the weekend.  Coming back it looks as though I've managed to derail this proposal again, which wasn't my intention at all—I actually support it.

It's a valid concern that websites might break when new permissions are specified.  I think there are solutions, though.  Suppose we started with `deny-all-v1`, which denies all permissions specified at the present time.  When the set of permissions changes, we create `deny-all-v2`, and so on.

Potentially there would end up being a lot of versions, but that's not a problem.  The browser just needs to know which "generation" each permission belongs to; this tells it the version of `deny-all` which is required for it to be disabled.  Web developers will normally just pick the latest version of `deny-all` for their project, since they want to deny everything which is specified at that time.

**Comment by vphantom** [comment]
There is nothing inherently wrong with a `deny-all`. Web sites routinely need updating to keep up with current standards as browsers evolve, and the use of `deny-all` itself being opt-in makes webmasters take responsibility explicitly for watching out for future issues.  Versions would bring us right back to having to discover and disable new versions proactively.

**Comment by kifd** [comment]
> ... since they want to deny everything which is specified at that time.

No, I want to deny everything that I've not explicitly asked for, regardless of if that permission exists now or in the future. I know what features the websites I serve up need, they're the ones I enable - I really don't want to continually look out for an ever increasing laundry list of permission features that need disabling.

I just can't see how a choice of "deny all except a,b,c features" would manage to break a site in the future, without the website itself changing. At that point I can simply add the needed "x,y,z features" to the directives whitelist and be thankful that I haven't had to waste my time disabling "d,e,f,g,h ... features" in the meantime.

Besides, as originally pointed out by OP six years ago, it's how the Content-Security-Policy works and nobody's proposing to change that to this state of affairs.

**Comment by clelland** [comment]
@annevk, I think that we can do this safely, and that something shaped roughly like the proposed

```http
Permissions-Policy: untrusted
```

is a reasonable way to do it.

The model I have is that there are a set of policy-controlled features which, by default, we will deny to cross-origin embedded content, on the basis that we don't trust it.  This applies to any cross-origin page loaded into an `<iframe>` with no `allow` attribute at all.

However, we expect that such embedded content will continue to fundamentally *work*. We wouldn't choose to put an existing feature under policy-control (like the above-mentioned strawmen 'images' or 'scripts') if we believed that it would break existing embedded content. Any addition like that could only happen after very careful consideration of backwards-compatibility and interop. That's already a promise we have to make, because of existing embedded content, not because of this proposal.

This proposal just allows sites to use a header directive to opt in to that same mode for their top-level documents.

The risk to the site should be roughly the same as being used as embedded content -- and is actually strictly less, since a site which can set its own headers can also mitigate any breakage by adjusting those headers, unlike a third-party embed.

**Comment by 0xVavaldi** [comment]
From what I can see a vast majority of people is for implementing some form of catch all. Be it a default deny or a default allow. Behaviour similar to CSP or something akin to it. Is there a specific reason why It's been 8 years and no one has started to implement this (to my knowledge)?

**Comment by ewanm89** [comment]
> From what I can see a vast majority of people is for implementing some form of catch all. Be it a default deny or a default allow. Behaviour similar to CSP or something akin to it. Is there a specific reason why It's been 8 years and no one has started to implement this (to my knowledge)?

We don't need default allow, no header at all, and currently not including a feature does that... We do need default deny, 1. for basic security practice of deny all unless approved, 2. so we don't have stupidly long headers from lots of features, there is already quite a few of them.

> I'm sorry! 😢 I posted what was meant to be a question about a minor detail, then went away for the weekend. Coming back it looks as though I've managed to derail this proposal again, which wasn't my intention at all—I actually support it.
> 
> It's a valid concern that websites might break when new permissions are specified. I think there are solutions, though. Suppose we started with `deny-all-v1`, which denies all permissions specified at the present time. When the set of permissions changes, we create `deny-all-v2`, and so on.
> 
> Potentially there would end up being a lot of versions, but that's not a problem. The browser just needs to know which "generation" each permission belongs to; this tells it the version of `deny-all` which is required for it to be disabled. Web developers will normally just pick the latest version of `deny-all` for their project, since they want to deny everything which is specified at that time.

As a developer/administrator I do not want to go in every time someone adds another feature and have to bump the version number, if I need some specific feature I can go in an specify just that feature overriding the default deny. This is really, really basic.

Let me be very clear, at no point should we be allowing something because a programmer/administrator didn't keep up with browser changes this violates basic security priniciple. If that means a browser update blocks something that breaks a website (something a lot more visible), I would fix this issue when it comes up. There seems to be some big worry that a website will suddenly break in some browser because they added a new permission for a feature that needs to be added to the header and allowed, this is not a problem. I would rather my bank, google search or facebook to be broken in a secure way for an hour blocking a load of potentially problematic features while someone goes in an makes a single line change to a config file for headers to send to add the one feature they need and then they can think of whether they really need that feature.

**Comment by abjugard** [comment]
> I would rather my bank, google search or facebook to be broken in a secure way for an hour blocking a load of potentially problematic features while someone goes in an makes a single line change to a config file for headers to send to add the one feature they need and then they can think of whether they really need that feature.

100%, and if someone is truly worried about this, they simply won't set the opt-in `default-deny` directive in their particular application.

**Comment by PeteX** [comment]
The problem is that the permission flags might affect things in two directions. A new flag might allow existing behaviour to be disabled, for example there might be a wish to restrict `fetch` on pages which don't use it. Alternatively, there might be a new API, and when it's introduced, it brings with it a new permission flag.

`deny-all` could reasonably restrict permission flags of the second type, but if it restricts flags of the first type, it risks suddenly turning `fetch` off for lots of existing websites when the `fetch` flag is introduced. (It would also be a new source of browser incompatibility, because browsers may adopt the new flag at different times.) My view is that this is unacceptable and would make the `deny-all` feature too dangerous to use, though possibly that is not everyone's opinion?

A design which avoids this problem isn't impossible. One option is the `deny-all-vX` flags I discussed earlier. It would also be possible to make `deny-all` disable only new APIs, though I feel that reduces the usefulness. Website authors may well wish to opt out of many standard web platform features, before then re-enabling the exact ones they need.

**Comment by abjugard** [comment]
> `deny-all` could reasonably restrict permission flags of the second type, but if it restricts flags of the first type, it risks suddenly turning `fetch` off for lots of existing websites when the `fetch` flag is introduced. (It would also be a new source of browser incompatibility, because browsers may adopt the new flag at different times.) My view is that this is unacceptable and would make the `deny-all` feature too dangerous to use, though possibly that is not everyone's opinion?

I think this is where the opinions differ, in my opinion it is fully reasonable to allow derelict unmaintained applications to silently break, it's better that they break than that they are insecure by default.

I expect apps to be maintained, and if they are then with clear communication when things are added to the spec (with a couple of months of warning until additions become enforcing), keeping the policy up to date isn't a problem, as long as people care, and if they don't care they won't have enabled any security to begin with.

During this period any newcomers to the spec can be included in a policy and work as intended, if they aren't declared and a site uses deny-all, the browser should warn about what would be a violation in the future (in a manner which lets client side logging frameworks catch it and report errors).

Alternatively, I could see an acceptable compromise here, where directives are put into groups e.g. anything related to fetch and other communication methods would be in the connectivity group, things like location/accelerometer/ambient-light-sensor could be called sensors and the directive could be declared as e.g. `default-deny=(connectivity sensors)`.

**Comment by PeteX** [comment]
Perhaps the only way round this is to support both options. For example, `deny-all` could do what @abjugard suggests, and `deny-all-vX` could do what I suggested in [my comment](https://github.com/w3c/webappsec-permissions-policy/issues/189#issuecomment-2424976297) some time ago.

**Comment by ewanm89** [comment]
> The problem is that the permission flags might affect things in two directions. A new flag might allow existing behaviour to be disabled, for example there might be a wish to restrict `fetch` on pages which don't use it. Alternatively, there might be a new API, and when it's introduced, it brings with it a new permission flag.
> 
> `deny-all` could reasonably restrict permission flags of the second type, but if it restricts flags of the first type, it risks suddenly turning `fetch` off for lots of existing websites when the `fetch` flag is introduced. (It would also be a new source of browser incompatibility, because browsers may adopt the new flag at different times.) My view is that this is unacceptable and would make the `deny-all` feature too dangerous to use, though possibly that is not everyone's opinion?
> 
> A design which avoids this problem isn't impossible. One option is the `deny-all-vX` flags I discussed earlier. It would also be possible to make `deny-all` disable only new APIs, though I feel that reduces the usefulness. Website authors may well wish to opt out of many standard web platform features, before then re-enabling the exact ones they need.

It should block fetch, I get report page x is broken in browser y, I generally develop on browser x, I take a look and see the error of the permission policy block in the console. I take a look at the browser x changelog and see they have added permissions to fetch, I know I need fetch on my site for it to work so I go edit the line from `default-deny` to `default-deny fetch=(self)` or `default-deny fetch=(src)` push to change, check it is now working and continue my day. There are numerous ways sites can break on a browser update, we can't ever guarantee every change will 100% be compatible with every site already. It happens. This is an easy one to diagnose and solve.

**Comment by ewanm89** [comment]
> > The problem is that the permission flags might affect things in two directions. A new flag might allow existing behaviour to be disabled, for example there might be a wish to restrict `fetch` on pages which don't use it. Alternatively, there might be a new API, and when it's introduced, it brings with it a new permission flag.
> > `deny-all` could reasonably restrict permission flags of the second type, but if it restricts flags of the first type, it risks suddenly turning `fetch` off for lots of existing websites when the `fetch` flag is introduced. (It would also be a new source of browser incompatibility, because browsers may adopt the new flag at different times.) My view is that this is unacceptable and would make the `deny-all` feature too dangerous to use, though possibly that is not everyone's opinion?
> > A design which avoids this problem isn't impossible. One option is the `deny-all-vX` flags I discussed earlier. It would also be possible to make `deny-all` disable only new APIs, though I feel that reduces the usefulness. Website authors may well wish to opt out of many standard web platform features, before then re-enabling the exact ones they need.
> 
> It should block fetch, I get report page x is broken in browser y, I generally develop on browser x, I take a look and see the error of the permission policy block in the console. I take a look at the browser x changelog and see they have added permissions to fetch, I know I need fetch on my site for it to work so I go edit the line from `default-deny` to `default-deny fetch=(self)` or `default-deny fetch=(src)` push to change, check it is now working and continue my day. There are numerous ways sites can break on a browser update, we can't ever guarantee every change will 100% be compatible with every site already. It happens. This is an easy one to diagnose and solve.

Meanwhile my basic static html continues working just fine with a `default-deny` being set.

**Comment by Gunni** [comment]
Are you serious?

I just used https://www.permissionspolicy.com/ to generate a deny everything policy, and then had to add report to for them all... Individually...

So here's my deny everything policy:

```
permissions-policy: accelerometer=();report-to=default, ambient-light-sensor=();report-to=default, autoplay=();report-to=default, battery=();report-to=default, camera=();report-to=default, cross-origin-isolated=();report-to=default, display-capture=();report-to=default, document-domain=();report-to=default, encrypted-media=();report-to=default, execution-while-not-rendered=();report-to=default, execution-while-out-of-viewport=();report-to=default, fullscreen=();report-to=default, geolocation=();report-to=default, gyroscope=();report-to=default, keyboard-map=();report-to=default, magnetometer=();report-to=default, microphone=();report-to=default, midi=();report-to=default, navigation-override=();report-to=default, payment=();report-to=default, picture-in-picture=();report-to=default, publickey-credentials-get=();report-to=default, screen-wake-lock=();report-to=default, sync-xhr=();report-to=default, usb=();report-to=default, web-share=();report-to=default, xr-spatial-tracking=();report-to=default, clipboard-read=();report-to=default, clipboard-write=();report-to=default, gamepad=();report-to=default, speaker-selection=();report-to=default, conversion-measurement=();report-to=default, focus-without-user-activation=();report-to=default, hid=();report-to=default, idle-detection=();report-to=default, interest-cohort=();report-to=default, serial=();report-to=default, sync-script=();report-to=default, trust-token-redemption=();report-to=default, unload=();report-to=default, window-placement=();report-to=default, vertical-scroll=();report-to=default
```

1500 bytes, just this header?

**Comment by vphantom** [comment]
Why `report-to` at all? The only mention of `report-to` I can find in the MDN docs is for the `Content-Security-Policy` header, not for this `Permissions-Policy`.

**Comment by rgant** [comment]
@vphantom The documentation probably does need to be updated, but you can see an example for the 
[`Reporting-Endpoints`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Reporting-Endpoints#specifying_multiple_reporting_endpoints) header setting the "permissions-endpoint" for "permissions-policy-reports".

It is further documented in the [Reporting API](https://developer.mozilla.org/en-US/docs/Web/API/Reporting_API#concepts_and_usage) section that this covers both CSP and Permissions-Policy violations.

**Comment by rgant** [comment]
I think that a lot of the concerns about deny-all option boils down to not trusting developers. I would default to assuming that a developer wanting to use Permissions Policy header would research the risks and rewards. If they don't, then you cannot really help them.

Why not just require `report-to` field to be included when using `deny-all`?

**Comment by RealAlphabet** [comment]
> I think that a lot of the concerns about deny-all option boils down to not trusting developers. I would default to assuming that a developer wanting to use Permissions Policy header would research the risks and rewards. If they don't, then you cannot really help them.
> 
> Why not just require `report-to` field to be included when using `deny-all`?

This has nothing to do with trust. It is about responsibility and defense in depth.

A developer may deliberately choose a strict deny-by-default policy precisely because they are responsible for limiting what their application is capable of doing to users — including when their own application or software supply chain is compromised.

And that risk is not theoretical. Modern websites depend on frameworks, libraries, transitive dependencies, CDNs, third-party scripts, and external infrastructure. A vulnerability or compromise anywhere in that chain may result in unintended code executing in the application's origin.

In fact, this is one of the purposes of Permissions Policy itself: limiting the capabilities available to both first-party and third-party content so that unexpected or compromised code has less authority.

If security depended on trusting every piece of code involved, there would be little point in capability restrictions in the first place.

Security controls should not rely solely on assumptions about human intentions. They should reduce risk whether the cause is malicious behavior, an implementation mistake, or a compromise, and limit the damage that can result.

Regarding `report-to`: reporting is certainly useful, but I see it as orthogonal to `deny-all`. Reporting helps developers detect violations; it does not replace the security property provided by deny-by-default.

If a developer explicitly opts into `deny-all`, they are also explicitly choosing the compatibility trade-off that comes with future policy-controlled features. That seems like a reasonable choice for the platform to allow developers to make.

**Comment by RealAlphabet** [comment]
There may also be a straightforward way to address the forward-compatibility concern.

When a browser introduces a new policy-controlled feature, `deny-all` does not necessarily have to start enforcing it immediately.

The browser could initially treat that new feature as report-only for documents using `deny-all`: allow its use for a defined compatibility period, but generate a Permissions Policy report indicating that the feature will eventually be denied unless the policy is updated.

During that transition period, developers would still remain fully in control. They could explicitly allow the new feature if their application requires it, or explicitly deny it immediately if they want the strict behavior from day one.

Conceptually:
- no explicit rule → temporarily allowed + reported for compatibility;
- explicit allow → allowed;
- explicit deny → blocked immediately.

After a reasonable number of browser releases, the compatibility exception would expire and the new feature would become subject to `deny-all` by default.

This seems to preserve all three desirable properties:
- existing applications are not unexpectedly broken when a browser ships a new policy-controlled feature;
- security-conscious applications can immediately deny the new capability;
- `deny-all` still provides meaningful future-proof deny-by-default semantics once the transition period ends.

Browsers already have a report-only mechanism for Permissions Policy, so this could potentially reuse existing infrastructure rather than weakening the long-term security semantics of `deny-all`.

**Comment by Dreamsorcerer** [comment]
Alternatively, you could version the spec and have deny-all-x.y or something, which denies everything defined at that version. Developers would need to update it occasionally if they want to ensure new features are denied, but it solves the issue of having a potentially MB long header of every single option enumerated.

**Comment by jvoisin** [comment]
> Alternatively, you could version the spec and have deny-all-x.y or something, which denies everything defined at that version. Developers would need to update it occasionally if they want to ensure new features are denied, but it solves the issue of having a potentially MB long header of every single option enumerated.

"Please regularly update this security control on your website by periodically looking at current web specifications" isn't a great selling point.

**Comment by Dreamsorcerer** [comment]
"Please regularly check every old website you ever wrote hasn't been broken by a browser update" isn't a great selling point either, and exactly what the maintainers have flagged as a concern... I see the argument both ways, and nobody has provided a solution that satisfies both sides. My suggestion just avoids the silly maintenance of a list of dozens or hundreds of individual options needed to be put into a header with the current design. It solves part of the problem, while the rest of the problem still appears unsolvable in a way that satisfies everyone.

**Comment by Gunni** [comment]
> "Please regularly update this security control on your website by periodically looking at current web specifications" isn't a great selling point.

The current paradigm is the same, except much worse because instead of one deny all you need hundreds?

**Comment by Dreamsorcerer** [comment]
> > "Please regularly update this security control on your website by periodically looking at current web specifications" isn't a great selling point.
> 
> The current paradigm is the same, except much worse because instead of one deny all you need hundreds?

Well, no. Your website doesn't break tomorrow when a new feature is added. The tradeoff today is that you potentially lose a degree of security tomorrow. The maintainers here clearly prefer the security tradeoff to having loads of websites broken when developers stop supporting them.

**Comment by Gunni** [comment]
> Well, no. Your website doesn't break tomorrow when a new feature is added. The tradeoff today is that you potentially lose a degree of security tomorrow. The maintainers here clearly prefer the security tradeoff to having loads of websites broken when developers stop supporting them.

The suggestion was a versioned deny all, how will that break anything, ever?

**Comment by ecki** [comment]
Even if we don’t do it Versionen, there are means to avoid breakage, one is the delayed default mentioned above and one is my favorite to define a semantic understanding what does not need permissions (ever). Like if you stay within those bounds you can be fairly safe that only future features might introduce permissions. And also it should degrade safely, so maybe asking for manual permission or just not,showing all content or animations.

Btw a little compromise, we can add default denies like “not-mobile#2026” which can bundle common properties like sensors, location, orientation, mic, … so I don’t have to list hindreds of usual denies for certain applications (think stupid office applications or public untrusted web sites)

**Comment by Dreamsorcerer** [comment]
> > Well, no. Your website doesn't break tomorrow when a new feature is added. The tradeoff today is that you potentially lose a degree of security tomorrow. The maintainers here clearly prefer the security tradeoff to having loads of websites broken when developers stop supporting them.
> 
> The suggestion was a versioned deny all, how will that break anything, ever?

Wait, I'm confused. My suggestion was a versioned deny all, I thought you were arguing against this. The suggestion before mine was to automatically add things to the deny all rule after a phase-in period, which would certainly break unmaintained websites in the long run.

**Comment by Gunni** [comment]
> Wait, I'm confused. My suggestion was a versioned deny all, I thought you were arguing against this. The suggestion before mine was to automatically add things to the deny all rule after a phase-in period, which would certainly break unmaintained websites in the long run.

To be absolutely clear. I'm in favor of versioned deny-all.

**Comment by RealAlphabet** [comment]
> > Wait, I'm confused. My suggestion was a versioned deny all, I thought you were arguing against this. The suggestion before mine was to automatically add things to the deny all rule after a phase-in period, which would certainly break unmaintained websites in the long run.
> 
> To be absolutely clear. I'm in favor of versioned deny-all.
> 

Versioned against what, exactly?

Permissions Policy does not currently define a single, versioned set of policy-controlled features. Each user agent has its own set of supported features, user agents are not required to support the same set, and policy-controlled features can be defined by other specifications or even be implementation-specific.

So a "deny-all-x.y" tied to a Permissions Policy spec version would seem to require introducing an additional normative, versioned registry mapping each version to a frozen set of features.

Otherwise, what does "x.y" actually identify?

If it refers to the Permissions Policy specification version, that version does not uniquely identify the set of features implemented by a browser. If it refers to the browser version, its meaning becomes vendor-specific. And if it refers only to standardized features, non-standard or experimental policy-controlled features are left out.

I think there may be a simpler way to preserve the same compatibility property while making the intended security semantics explicit.

Instead of versioning the specification, "deny-all" could either be strict, or carry a date/epoch cutoff.

Conceptually:

- "deny-all=strict"
  
  - Deny every policy-controlled feature supported by the user agent unless explicitly allowed.
  - This includes features that become policy-controlled in the future.

- "deny-all=2026-09"
  
  - Deny every supported feature that became policy-controlled on or before September 2026.
  - Features that become policy-controlled after that cutoff retain their normal behavior unless explicitly denied.
  - If one of those newer features is actually used, the browser could generate a Permissions Policy report informing the developer that the current deny-all snapshot does not cover it.

Explicit per-feature rules would take precedence in either direction.

An important detail is that the relevant date would be when a feature became policy-controlled, not necessarily when the underlying API was originally introduced.

For example, suppose an API has existed since 2024 but browsers add a new Permissions Policy control for it in 2027. A site using "deny-all=2026-09" would not suddenly break in 2027. The API would remain usable, but its use could generate a report telling the developer that a new policy-controlled capability exists outside their selected cutoff.

A site that really wants future-proof deny-by-default semantics can instead use "strict" and deliberately accept that trade-off.

I originally suggested that browsers could provide a compatibility/grace period before new features became subject to "deny-all", but with a dated cutoff I don't think that is necessary anymore.

The developer effectively chooses the compatibility boundary themselves.

The browser no longer needs to keep track of grace periods, release counts, or when enforcement should automatically transition. It only needs stable metadata indicating when each supported feature became policy-controlled. Existing Permissions Policy reporting infrastructure could then provide the migration signal when a feature newer than the application's cutoff is actually exercised.

This also gives unmaintained applications the behavior that a versioned deny-all was trying to preserve: their security boundary remains frozen rather than unexpectedly breaking later.

But it does so without pretending that there is a universal "Permissions Policy 1.x" feature set shared by every browser.

For applications where that frozen boundary is undesirable, "strict" retains the actual future-proof meaning of “deny everything unless I explicitly allow it.”

This seems to address the forward-compatibility concern without requiring a globally versioned feature registry.

**Comment by RealAlphabet** [comment]
P.S. One portability issue occurred to me after posting this: the cutoff probably should not be based on the date a particular user agent implemented a policy-controlled feature.

Otherwise, "deny-all=2026-09" could mean different things across browsers. A feature introduced as a policy in Firefox in 2025 but only in Chrome in 2027 could be denied in one and allowed in the other, even though the author selected the same cutoff.

The cutoff would therefore need to refer to a normative introduction epoch associated with the policy-controlled feature itself, independent of individual browser implementation dates.

In other words, the feature gets a stable epoch when it enters the standardized Permissions Policy surface, and that epoch must not be backdated based on earlier vendor-specific implementations.

Unregistered/vendor-specific features could simply remain outside dated cutoffs, while "strict" would still cover everything supported by the user agent.

That would preserve the important property that a dated "deny-all" has the same normative meaning across conforming user agents.

**Comment by Gunni** [comment]
> Permissions Policy does not currently define a single, versioned set of policy-controlled features. Each user agent has its own set of supported features, user agents are not required to support the same set, and policy-controlled features can be defined by other specifications or even be implementation-specific.
> 
> So a "deny-all-x.y" tied to a Permissions Policy spec version would seem to require introducing an additional normative, versioned registry mapping each version to a frozen set of features.

I would argue that the lack of a canonical list of feature flags and versions is one of the worst features of this spec and is worth discussing.

It is always going to mean different behavior for different browsers.

Is this one of the reasons that if you try to do a deny everything today some browsers will start warning you that you're denying or controlling things that don't exist to them.

I propose defining versioning semantics and that this spec contain canonical lists of features to control.

Maybe we should move that to a new discussion though, but in my opinion, default deny depends on it.

**Comment by RealAlphabet** [comment]
I agree that a canonical registry becomes necessary for the portable dated cutoff.

For `deny-all=YYYY-MM` to have the same meaning across user agents, the specification would need a canonical list of standardized policy-controlled features, with a normative introduction epoch associated with each feature.

I would make the distinction explicit:

- `deny-all=strict`
  - Applies to every policy-controlled feature supported by the current user agent, including non-standard and implementation-specific features.

- `deny-all=YYYY-MM`
  - Applies only to standardized policy-controlled features whose normative introduction epoch is less than or equal to the specified cutoff.
  - Non-standard or implementation-specific features MUST NOT be silently included in this baseline, since doing so would make the same value mean different things across user agents.

- `deny-all=YYYY-MM-vendor`
  - An optional, explicitly user-agent-specific snapshot.
  - User agents that expose non-standard policy-controlled features **SHOULD** use this form if they want to provide a dated deny-all baseline that includes those features.
  - Such a snapshot may include both the standardized baseline up to `YYYY-MM` and implementation-specific policy-controlled features exposed by that user agent up to that date.

For example:

```http
deny-all=2026-09
```

would have a portable meaning defined entirely by the standardized registry.

Whereas:

deny-all=2026-09-webkit

could explicitly mean:

> the standardized 2026-09 baseline, plus WebKit-specific policy-controlled features included in WebKit's own September 2026 snapshot.

The important property is that implementation-specific features never silently change the meaning of the unsuffixed YYYY-MM form.

If another user agent does not recognize the vendor-qualified value, it should treat it as an unknown implementation-specific value rather than attempting to reinterpret it as the portable baseline.

This also means that the current console-warning behavior for unknown feature identifiers is a separate issue. The Permissions Policy processing model can continue to allow different user agents to support different subsets, while the canonical registry is only needed to give the portable dated cutoff a stable meaning.

So I would separate the two concerns:

1. a normative registry + epochs for the interoperable YYYY-MM baseline;


2. an optional YYYY-MM-vendor namespace for implementation-specific extensions.

That gives vendors room to expose their own additional policy-controlled features without making the standardized cutoff dependent on browser-specific history.

**Comment by Dreamsorcerer** [comment]
Yeah, I would expect that if such a rule was added to the spec, then it would define an explicit list of features at that point in time. That could be set of features that already have wide support across browsers at that time, for example. With other features being listed as experimental or similar and not included in the deny list until a later revision.
